### ISSUE TITLE
Training poor after adding single 2D raycast observation 

### ISSUE THREAD

**Issue by Zoturi** [Unity-Technologies/ml-agents#2305]
Hi. I have a basic 2D top down level, where the agent is controlling an actor which tries to find the exit door. This doors location is randomized to be one of three possibilities. In the most basic setup, there is only single normalized vector added as observation and the agent is rewarded 1.0f and set to done, when it finds the door. This can be trained most of the times successfully, not always though, but I guess that is due to the algorithm and luck, if the agent happens to find the door early and learn the meaning of the directional vector. 

My issue is, that when I add a single 2D ray cast as an observation, like this:
`float rayDistance = 30f;
float[] rayAngles = {0f, 90f, 180f, 270f, 30f, 60f, 120f, 150f, 210f, 240f, 300f, 330f};
string[] detectableObjects = {"DZ_wall"};
AddVectorObs(agentRayPer2D.Perceive(rayDistance,rayAngles,detectableObjects));`

and apply the tag to the walls, but keep everything otherwise the same, the training slows down significantly. Is this to expected or is this not normal?

**Comment by ervteng** [comment]
Yes, it's expected - you have an observation space that's 12 times larger than before. And a neural network that's larger as well.

**Comment by DuracellTurtle** [comment]
Actually, if you look into the RayPerception, the resulting vector's length is rayAngles * (detectableObjects + 2) so you went from 1 input to 37. Having only one detectable object, or if it is not important which one you hit, you could reduce the size since you don't need the one-hot value(s) describing what object was hit. Last thing to note, angle 0 goes towards positive x (right)

**Comment by Zoturi** [comment]
> Actually, if you look into the RayPerception, the resulting vector's length is rayAngles * (detectableObjects + 2) so you went from 1 input to 37. Having only one detectable object, or if it is not important which one you hit, you could reduce the size since you don't need the one-hot value(s) describing what object was hit. Last thing to note, angle 0 goes towards positive x (right)

Could you elaborate a bit more on this please. I'm not sure I know what you mean with the "one-hot values". My goal is to have multiple separate tags, walls, hazards (pits for the actor to fall in, and spikes to dodge by rolling), collectibles and enemies. So a rather healthy list of things to recognize, with different actions to take.  

So I would need to optimize the amount, to make the training more efficient. One thing of course it to have as few rays as possible. Thanks for the help.

**Comment by DuracellTurtle** [comment]
Generally, one hot encoding means that to encode N possible values, you don't use one number from 0 to N-1, but and encode it as N numbers with one of them being 1 and the rest 0. From what I understand, this is quite necessary for the neural network to be able to understand this kind of discrete information.
That's what the RayPerception does for each ray, plus a value that's set to 1 if no object is hit, plus another value for the distance, if an object is hit. If you will have multiple types of objects that it should react differently to, then you need it and all you can do is balance the number of rays and size of the network

**Comment by DuracellTurtle** [comment]
Also you could group some tags to the same category if it makes sense to react to them the same way, you would need to adjust the RayPerception yourself in that case
